README.md
MGEvents
========

Blocks based custom event and UIControl event handlers


CHANGELOG.markdown
## 1.1.0

- Added support for watchOS 2 (@staminajim)
- Add a way to listen for events from any object of a specified class 
  (@staminajim)
- Fixed a bug with [when: does: doWithContext:] (@staminajim)
- Added the ability to perform actions "onAnyOf" a list of triggers 
  (@staminajim)
- Renamed block declaractions to avoid clashes (@edwardvalentini)
- Added the ability to watch for changes on multiple key paths with one action 
  block (@staminajim)

## 1.0.0

Initial release.


MGEvents/NSObject+MGEvents.h
//
//  Created by matt on 24/09/12.
//

#import "MGBlockWrapper.h"

/**
* Provides lightweight, blocks based keypath observing and custom events.
*/

@interface NSObject (MGEvents)

/** @name Custom event observing */

@property (nonatomic, retain) NSMutableDictionary *MGEventHandlers;

/**
When a custom event is triggered with the given name, perform the given block.

    [earth on:@"ChangedShape" do:^{
        NSLog(@"the earth has changed shape");
    }];
*/
- (void)on:(NSString *)eventName do:(MGBlock)handler;

/**
 When a custom event is triggered by any of the given names, perform the given block.

 [earth onAnyOf:@[@"ChangedShape", @"StoppedSpinning"] do:^{
     NSLog(@"the earth has changed shape or stopped spinning");
 }];
 */
- (void)onAnyOf:(NSArray *)eventNames do:(MGBlock)handler;

/**
* When a custom event is triggered with the given name, perform the given block
* once. The block will not be performed on future triggering of the same event.
*/
- (void)on:(NSString *)eventName doOnce:(MGBlock)handler;

/**
When a custom event is triggered with the given name, perform the given block.
The block may potentially be provided a context object.

    [earth on:@"ChangedShape" doWithContext:^(id context) {
        NSLog(@"the earth has changed shape.");
        NSLog(@"some details about the shape change: %@", context);
    }];
*/
- (void)on:(NSString *)eventName doWithContext:(MGBlockWithContext)handler;

/**
 When a custom event is triggered with any of the given names, perform the given block.
 The block may potentially be provided a context object.

 [earth onAnyOf:@[@"ChangedShape", @"StoppedSpinning"] doWithContext:^(id context) {
     NSLog(@"the earth has changed shape or stopped spinning.");
     NSLog(@"some details about the change: %@", context);
 }];
 */
- (void)onAnyOf:(NSArray *)eventNames doWithContext:(MGBlockWithContext)handler;

/**
 When a particular object triggers the specified event, perform the given block.

 [self when:earth does:@"ChangedShape" do:^{
    NSLog(@"the earth has changed shape");
 }];
*/

- (void)when:(id)object does:(NSString *)eventName do:(MGBlock)handler;

/**
 When a particular object triggeres any of the specified events, perform the given block.

 [self when:earth doesAnyOf:@[@"ChangedShape", @"StoppedSpinning"] do:^{
     NSLog(@"the earth has changed shape or stopped spinning");
 }];
 */

- (void)when:(id)object doesAnyOf:(NSArray *)eventNames do:(MGBlock)handler;

/**
 When a particular object triggers the specified event, perform the given block.
 The block may potentially be provided a context object.
 [self when:earth does:@"ChangedShape" doWithContext:^(id context) {
     NSLog(@"some details about the change: %@", context);
 }];
*/

- (void)when:(id)object does:(NSString *)eventName doWithContext:(MGBlockWithContext)handler;

/**
 When a particular object triggers any of the specified events, perform the given block.
 The block may potentially be provided a context object.
 [self when:earth doesAnyOf:@[@"ChangedShape", @"StoppedSpinning"] doWithContext:^(id context) {
 NSLog(@"some details about the change: %@", context);
 }];
 */

- (void)when:(id)object doesAnyOf:(NSArray *)eventNames doWithContext:(MGBlockWithContext)handler;


/** @name Custom event triggering */

/**
Trigger a custom event.

    [earth trigger:@"ChangedShape"];
*/
- (void)trigger:(NSString *)eventName;

/**
Trigger a custom event, providing context.

    [earth trigger:@"ChangedShape" withContext:@{ @"newShape": @"flat" }];
*/
- (void)trigger:(NSString *)eventName withContext:(id)context;

/** @name Keypath observing */

@property (nonatomic, retain) NSMutableDictionary *MGObservers;

/**
On change of the given keypath, perform the given block.

    [box onChangeOf:@"selected" do:^{
        NSLog(@"my selected state changed to: %@", box.selected ? @"ON" : @"OFF");
    }];
*/
- (void)onChangeOf:(NSString *)keypath do:(MGBlock)block;

/**
On change of any of the given keypaths, perform the given block.

    [box onChangeOfAny:@[@"selected", @"highlighted"] do:^{
        NSLog(@"my selected state is: %@", box.selected ? @"ON" : @"OFF");
        NSLog(@"my highlighted state is: %@", box.highlighted ? @"ON" : @"OFF");
    }];
*/
- (void)onChangeOfAny:(NSArray *)keypaths do:(MGBlock)block;

@property (nonatomic, copy) MGBlock onDealloc;

@end


MGEvents/UIControl+MGEvents.m
//
//  Created by matt on 19/10/12.
//

#import "UIControl+MGEvents.h"
#import <objc/runtime.h>

static char *MGEventHandlersKey = "MGEventHandlersKey";

@implementation UIControl (MGEvents)

- (void)onControlEvent:(UIControlEvents)controlEvent do:(MGBlock)handler {

  // get all handlers for this control event
  NSMutableArray *handlers = self.MGEventHandlers[@((int)controlEvent)];
  if (!handlers) {
    handlers = @[].mutableCopy;
    self.MGEventHandlers[@((int)controlEvent)] = handlers;
  }

  // add the handler
  MGBlockWrapper *wrapper = [MGBlockWrapper wrapperForBlock:handler];
  [self addTarget:wrapper action:@selector(doit) forControlEvents:controlEvent];
  [handlers addObject:wrapper];
}

- (void)removeHandlersForControlEvent:(UIControlEvents)controlEvent {

  // get all handlers for this control event
  NSMutableArray *handlers = self.MGEventHandlers[@((int)controlEvent)];

  // remove them
  for (MGBlockWrapper *wrapper in handlers) {
    [self removeTarget:wrapper action:NULL forControlEvents:controlEvent];
  }

  // and now release the blocks
  handlers = @[].mutableCopy;
  self.MGEventHandlers[@((int)controlEvent)] = handlers;
}

#pragma mark - Getters and setters

- (NSMutableDictionary *)MGEventHandlers {
  id handlers = objc_getAssociatedObject(self, MGEventHandlersKey);
  if (!handlers) {
    handlers = @{ }.mutableCopy;
    self.MGEventHandlers = handlers;
  }
  return handlers;
}

- (void)setMGEventHandlers:(NSMutableDictionary *)handlers {
  objc_setAssociatedObject(self, MGEventHandlersKey, handlers,
      OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

@end


MGEvents/NSObject+MGEvents.m
//
//  Created by matt on 24/09/12.
//

#import "NSObject+MGEvents.h"
#import "MGObserver.h"
#import "MGWeakHandler.h"
#import "MGDeallocAction.h"
#import <objc/runtime.h>

static char *MGObserversKey = "MGObserversKey";
static char *MGEventHandlersKey = "MGEventHandlersKey";
static char *MGDeallocActionKey = "MGDeallocActionKey";

#define MGProxyHandlers @"MGProxyHandlers"

@implementation NSObject (MGEvents)

#pragma mark - Custom events

- (void)on:(NSString *)eventName do:(MGBlock)handler {
  [self on:eventName do:handler once:NO context:NO];
}

- (void)onAnyOf:(NSArray *)eventNames do:(MGBlock)handler {
    for (NSString *eventName in eventNames) {
        [self on:eventName do:handler];
    }
}

- (void)on:(NSString *)eventName doOnce:(MGBlock)handler {
  [self on:eventName do:handler once:YES context:NO];
}

- (void)on:(NSString *)eventName doWithContext:(MGBlockWithContext)handler {
  [self on:eventName do:(MGBlock)handler once:NO context:YES];
}

- (void)onAnyOf:(NSArray *)eventNames doWithContext:(MGBlockWithContext)handler {
    for (NSString *eventName in eventNames) {
        [self on:eventName doWithContext:handler];
    }
}

- (void)on:(NSString *)eventName do:(MGBlock)handler once:(BOOL)once
   context:(BOOL)isContextBlock {

  // get all handlers for this event type
  NSMutableArray *handlers = self.MGEventHandlers[eventName];
  if (!handlers) {
    handlers = @[].mutableCopy;
    self.MGEventHandlers[eventName] = handlers;
  }

  if (isContextBlock) {
    [handlers addObject:@{@"blockWithContext" : [handler copy], @"once" : @(once)
    }];
  } else {
    [handlers addObject:@{@"block" : [handler copy], @"once" : @(once)}];
  }
}

- (void)when:(id)object does:(NSString *)eventName do:(MGBlock)handler {
    [self when:object does:eventName do:handler context:NO];
}

- (void)when:(id)object doesAnyOf:(NSArray *)eventNames do:(MGBlock)handler {
    for (NSString *eventName in eventNames) {
        [self when:object does:eventName do:handler];
    }
}

- (void)when:(id)object does:(NSString *)eventName doWithContext:(MGBlockWithContext)handler {
    [self when:object does:eventName do:(MGBlock)handler context:YES];
}

- (void)when:(id)object doesAnyOf:(NSArray *)eventNames doWithContext:(MGBlockWithContext)handler {
    for (NSString *eventName in eventNames) {
        [self when:object does:eventName doWithContext:handler];
    }
}

- (void)when:(NSObject *)object does:(NSString *)eventName do:(MGBlock)handler
      context:(BOOL)isContextBlock {

    // get the proxy handlers array
    NSMutableArray *handlers = self.MGEventHandlers[MGProxyHandlers];
    if (!handlers) {
        handlers = @[].mutableCopy;
        self.MGEventHandlers[MGProxyHandlers] = handlers;
    }

    // get the target object's handlers array
    NSMutableArray *theirHandlers = object.MGEventHandlers[eventName];
    if (!theirHandlers) {
        theirHandlers = @[].mutableCopy;
        object.MGEventHandlers[eventName] = theirHandlers;
    }

    id handlerDict = isContextBlock
          ? @{@"blockWithContext":[handler copy], @"once":@NO}
          : @{@"block":[handler copy], @"once":@NO};

    // store the handler dict locally
    [handlers addObject:handlerDict];

    // give the target object a weak reference to it, so that if we dealloc
    // it goes away with us, instead of lingering forever
    [theirHandlers addObject:[MGWeakHandler handlerWithDict:handlerDict]];
}

- (void)whenAny:(Class)objectOfClass does:(NSString *)eventName do:(MGBlock)handler {
    [self when:objectOfClass does:[NSObject globalMGEventNameFor:eventName] do:handler];
}

- (void)whenAny:(Class)objectOfClass doesAnyOf:(NSArray *)eventNames do:(MGBlock)handler {
    NSMutableArray *mangledEventNames = NSMutableArray.new;
    for (NSString *eventName in eventNames) {
        [mangledEventNames addObject:[NSObject globalMGEventNameFor:eventName]];
    }
    [self when:objectOfClass doesAnyOf:mangledEventNames.copy do:handler];
}

- (void)whenAny:(Class)objectOfClass does:(NSString *)eventName doWithContext:(MGBlockWithContext)handler {
    [self when:objectOfClass does:[NSObject globalMGEventNameFor:eventName] doWithContext:handler];
}

- (void)whenAny:(Class)objectOfClass doesAnyOf:(NSArray *)eventNames doWithContext:(MGBlockWithContext)handler {
    NSMutableArray *mangledEventNames = NSMutableArray.new;
    for (NSString *eventName in eventNames) {
        [mangledEventNames addObject:[NSObject globalMGEventNameFor:eventName]];
    }
    [self when:objectOfClass doesAnyOf:mangledEventNames.copy doWithContext:handler];
}

- (void)trigger:(NSString *)eventName {
    [self trigger:eventName withContext:nil];
    if (!class_isMetaClass(object_getClass(self))) {
        [self.class trigger:[NSObject globalMGEventNameFor:eventName]];
    }
}

- (void)trigger:(NSString *)eventName withContext:(id)context {
  NSMutableArray *handlers = self.MGEventHandlers[eventName];
    for (id handler in handlers.copy) {
        NSDictionary *handlerDict = [handler isKindOfClass:MGWeakHandler.class]
              ? [handler dict]
              : handler;
        if (!handlerDict) {
            [handlers removeObject:handler];
            continue;
        }
        if (handlerDict[@"blockWithContext"]) {
            MGBlockWithContext block = handlerDict[@"blockWithContext"];
            block(context);
        } else if (handlerDict[@"block"]) {
            MGBlock block = handlerDict[@"block"];
            block();
        }
        if ([handlerDict[@"once"] boolValue]) {
            [handlers removeObject:handler];
        }
    }
    if (!class_isMetaClass(object_getClass(self))) {
        [self.class trigger:[NSObject globalMGEventNameFor:eventName] withContext:context];
    }
}

#pragma mark - Property observing

- (void)onChangeOf:(NSString *)keypath do:(MGBlock)block {

  // get observers for this keypath
  NSMutableArray *observers = self.MGObservers[keypath];
  if (!observers) {
    observers = @[].mutableCopy;
    self.MGObservers[keypath] = observers;

  }

  // make and store an observer
  MGObserver *observer = [MGObserver observerFor:self keypath:keypath block:block];
  [observers addObject:observer];

  __unsafe_unretained id _self = self;
  __unsafe_unretained id _observer = observer;
  observer.onDealloc = ^{
      [_self removeObserver:_observer forKeyPath:keypath];
  };
}

- (void)onChangeOfAny:(NSArray *)keypaths do:(MGBlock)block {
  for (NSString *keypath in keypaths) {
    [self onChangeOf:keypath do:block];
  }
}

#pragma mark - Getters

- (NSMutableDictionary *)MGEventHandlers {
  id handlers = objc_getAssociatedObject(self, MGEventHandlersKey);
  if (!handlers) {
    handlers = @{}.mutableCopy;
    self.MGEventHandlers = handlers;
  }
  return handlers;
}

- (NSMutableDictionary *)MGObservers {
  id observers = objc_getAssociatedObject(self, MGObserversKey);
  if (!observers) {
    observers = @{}.mutableCopy;
    self.MGObservers = observers;
  }
  return observers;
}

- (MGBlock)onDealloc {
  MGDeallocAction *wrapper = objc_getAssociatedObject(self, MGDeallocActionKey);
  return wrapper.block;
}

+ (NSString *)globalMGEventNameFor:(NSString *)objectEvent {
    return [objectEvent stringByAppendingString:@"-MGGlobalEvent"];
}

#pragma mark - Setters

- (void)setMGEventHandlers:(NSMutableDictionary *)handlers {
  objc_setAssociatedObject(self, MGEventHandlersKey, handlers,
      OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (void)setMGObservers:(NSMutableDictionary *)observers {
  objc_setAssociatedObject(self, MGObserversKey, observers,
      OBJC_ASSOCIATION_RETAIN_NONATOMIC);
}

- (void)setOnDealloc:(MGBlock)block {
  MGDeallocAction *wrapper = objc_getAssociatedObject(self, MGDeallocActionKey);
  if (!wrapper) {
    wrapper = MGDeallocAction.new;
    objc_setAssociatedObject(self, MGDeallocActionKey, wrapper,
        OBJC_ASSOCIATION_RETAIN_NONATOMIC);
  }
  wrapper.block = block;
}

@end
